pretyping/recordops.mli

```diff
@@ -72,30 +73,64 @@ type cs_pattern =
   | Sort_cs of Sorts.family
   | Default_cs
 
-type obj_typ = {
-  o_ORIGIN : GlobRef.t;
-  o_DEF : constr;
-  o_CTX : Univ.AUContext.t;
-  o_INJ : int option;      (** position of trivial argument *)
-  o_TABS : constr list;    (** ordered *)
-  o_TPARAMS : constr list; (** ordered *)
-  o_NPARAMS : int;
-  o_TCOMPS : constr list } (** ordered *)
+val eq_cs_pattern : Environ.env -> cs_pattern -> cs_pattern -> bool
+val pr_cs_pattern : cs_pattern -> Pp.t
 
 (** Return the form of the component of a canonical structure *)
 val cs_pattern_of_constr : Environ.env -> constr -> cs_pattern * int option * constr list
 
-val pr_cs_pattern : cs_pattern -> Pp.t
+type cs
 
-type cs = GlobRef.t * inductive
+val check_and_decompose_canonical_structure :
+  Environ.env -> Evd.evar_map -> GlobRef.t -> cs
+
+val register_canonical_structure :
+  warn:bool -> Environ.env -> Evd.evar_map -> cs -> unit
 
-val lookup_canonical_conversion : Environ.env -> (GlobRef.t * cs_pattern) -> constr * obj_typ
-val register_canonical_structure : warn:bool -> Environ.env -> Evd.evar_map ->
-  cs -> unit
 val subst_canonical_structure : Mod_subst.substitution -> cs -> cs
+val solution_of_canonical_structure : cs -> GlobRef.t
+
+(** The canonical solution of a problem (proj,val) is a global
+    [constant = fun abs : abstractions_ty => body] and
+    [body = RecodConstructor params canon_values] and the canonical value
+    corresponding to val is [val cvalue_arguments].
+    It is possible that val is one of the [abs] abstractions, eg [Default_cs],
+    and in that case [cvalue_abstraction = Some i] *)
+type canonical_solution = {
+  constant : EConstr.t;
+
+  abstractions_ty : EConstr.t list;
+  body : EConstr.t;
+
+  nparams : int;
+  params : EConstr.t list;
+  cvalue_abstraction : int option;
+  cvalue_arguments : EConstr.t list;
+}
+
```

Comment:
It could be convenient to wrap these record types in a module `CanonicalSolution.t` as to have better scoping of fields [and general advantages of the programming coding convention]

Reply:
It makes sense, but then I may as well rename the file. I mean, canonicalStructures.ml looks more sensible that recordops.ml and then I can just say `Solution.t`.

Reply:
As you prefer, if you think renaming makes sense that's ok to me too [tho the module name seems a bit long, would be nice if we could find a more compact one] Not a big deal in any case.

Reply:
I like to make short aliases, and IIRC recent ocaml is efficient at that.
I mean, instead of `open CanonicalStructures` I would `module CS = CanonicalStructures` and use `CS.foo` all over the place.